common/stage.py

```diff
@@ -460,7 +479,188 @@ def seed_context( self, kll_file ):
 		'''
 		kll_file.context.initial_context( kll_file.lines, kll_file.data, kll_file )
 
-	def process( self ):
+	def process_connect_ids(self, kll_file, apply_offsets):
+		lines = kll_file.data.splitlines()
+
+		# Basic Tokens Spec
+		# TODO Storing these somewhere central might be a reasonable idea
+		spec = [
+			( 'ScanCode', ( r'S((0x[0-9a-fA-F]+)|([0-9]+))', ) ),
+			( 'Operator', ( r'=>|<=|i:\+|i:-|i::|i:|:\+|:-|::|:|=', ) ),
+			( 'USBCode', ( r'U(("[^"]+")|(0x[0-9a-fA-F]+)|([0-9]+))', ) ),
+			( 'NumberBase10', ( r'(([1-9][0-9]*))', ) ),
+			( 'Number', ( r'-?(0x[0-9a-fA-F]+)|(0|([1-9][0-9]*))', ) ),
+			( 'Name', ( r'[A-Za-z_][A-Za-z_0-9]*', ) ),
+			( 'Misc', ( r'.', ) ),  # Everything else
+		]
+
+		# Tokens to filter out of the token stream
+		# useless = [ 'Space', 'Comment' ]
+		useless = ['Misc']
+
+		# Build tokenizer that appends unknown characters to Misc Token groups
+		# NOTE: This is technically slower processing wise, but allows for multi-stage tokenization
+		# Which in turn allows for parsing and tokenization rules to be simplified
+		tokenizer = make_tokenizer(spec)
+
+		try:
+			most_recent_connect_id = 0
+			most_recent_offset = 0
+			processed_lines = []
+			for line in lines:
+				tokens = [x for x in tokenizer(line) if x.type not in useless]
+
+				for r_element, mid_element, l_element in zip(tokens[0::3], tokens[1::3], tokens[2::3]):
+
+					# Preprocessor tag for offsetting the scancodes by a fixed amount
+					# This makes it eay to apply offsets to older files
+					# The scope for this term is for the current file
+					if (r_element.value == "ScanCodeOffset" and
+								mid_element.value == "=" and
+								l_element.type == "NumberBase10"):
+						most_recent_offset = int(l_element.value)
+
+					# Preprocessor definition for the connectId
+					# TODO these should likely be defined in their own file somewhere else
```

Comment:
Same as the previous comment, there is a danger of divergence as the definitions are hardcoded.

Reply:
Yeah, but I think it'll be ok. With probably the exception of the Operator.

Regardless, at this stage I think it's all good.